File: china_finance/public/js/month_end.js
frappe.provide("china_finance.month_end");

(() => {
	const call = async (method, name) => (await frappe.call({method: `china_finance.services.month_end.${method}`, args: {name}, type: "POST"})).message;
	const esc = value => frappe.utils.escape_html(String(value ?? ""));
	const labels = {Posting: "月末记账", Checking: "月末检查", Closing: "结转损益", Ready: "待确认结账", Paused: "已暂停", Failed: "待处理问题"};
	const show = (frm, result) => {
		frm.set_intro(`${esc(labels[result.state] || result.state)} · ${result.completed || 0}/${result.total || 0} 张<br>${esc(result.message)}`, ["Failed", "Paused"].includes(result.state) ? "orange" : "blue");
	};
	const run = async frm => {
		if (frm.__month_running) return;
		frm.__month_running = true;
		frm.__month_stop = false;
		try {
			if (frm.is_new() || frm.is_dirty()) await frm.save();
			let result = await call("start", frm.doc.name);
			show(frm, result);
			frm.add_custom_button(__("暂停"), async () => {
				frm.__month_stop = true;
				await call("pause", frm.doc.name);
				await frm.reload_doc();
			});
			while (!frm.__month_stop && ["Posting", "Checking", "Closing"].includes(result.state)) {
				result = await call("advance", frm.doc.name);
				show(frm, result);
				if (result.state === "Closing") await new Promise(resolve => setTimeout(resolve, 2000));
			}
		} finally {
			frm.__month_running = false;
			await frm.reload_doc();
		}
	};
	china_finance.month_end.render = frm => {
		for (const field of ["naming_series", "status", "previous_frozen_date", "closed_by", "closed_on", "reopened_by", "reopened_on", "reopen_reason", "checks_section", "checks", "notes", "preparation_state", "preparation_message"]) {
			frm.toggle_display(field, false);
		}
		frm.toggle_display("archive_package", !!frm.doc.archive_package);
		frm.toggle_display("period_closing_voucher", !!frm.doc.period_closing_voucher);
		frm.set_df_property("period_closing_voucher", "read_only", 1);
		const data = JSON.parse(frm.doc.preparation_data || "{}");
		if (frm.doc.preparation_state) show(frm, {state: frm.doc.preparation_state, message: frm.doc.preparation_message, completed: data.done?.length, total: data.items?.length});
		frm.fields_dict.preparation_preview.$wrapper.html(`<p>系统会完成本期凭证记账、月末检查和损益结转；核对财务报表后确认结账。关闭页面不会丢失进度。</p>${(data.checks || []).length ? `<table class="table table-bordered"><thead><tr><th>结账检查</th><th>结果</th><th>说明</th></tr></thead><tbody>${data.checks.map(r => `<tr><td>${esc(r.description)}</td><td>${r.passed ? "通过" : esc(r.severity === "Blocking" ? "待处理" : "提示")}</td><td>${esc(r.details)}</td></tr>`).join("")}</tbody></table>` : ""}`);
		frm.add_custom_button(__("本期凭证"), () => {
			frappe.route_options = {company: frm.doc.company, from_date: frm.doc.from_date, to_date: frm.doc.to_date, period_preset: "", voucher_status: "全部有效凭证"};
			frappe.set_route("query-report", "China Voucher Ledger");
		});
		frm.add_custom_button(__("试算平衡"), () => frappe.require("/assets/china_finance/js/voucher_preparation.js", () => china_finance.preparation.trial(frm.doc.company, frm.doc.from_date, frm.doc.to_date)));
		frm.add_custom_button(__("财务报表"), () => {
			frappe.route_options = {company: frm.doc.company, from_date: frm.doc.from_date, to_date: frm.doc.to_date};
			frappe.set_route("query-report", "China Financial Statements");
		});
		frm.add_custom_button(__("现金流量"), () => frappe.set_route("List", "China Cash Flow Assignment", {company: frm.doc.company}));
		if (frm.doc.docstatus === 1) {
			frm.set_intro(frm.doc.archive_package ? __("已结账，结账档案已生成") : __("已结账，结账档案正在生成，可稍后刷新"), "green");
			return;
		}
		if (frm.doc.preparation_state === "Ready") {
			frm.page.set_primary_action(__("确认结账"), () => frappe.confirm(__("已核对财务报表，确认结账并生成结账档案？"), async () => {
				await call("finish", frm.doc.name); await frm.reload_doc();
			}));
		} else {
			frm.page.set_primary_action(__(frm.doc.preparation_state ? "继续月末结账" : "开始月末结账"), () => frappe.confirm(__("系统将按日期记账本期已核对凭证，并完成月末检查和损益结转。确认开始？"), () => run(frm)));
		}
		if (["Posting", "Checking", "Closing"].includes(frm.doc.preparation_state) || frm.__month_running) {
			frm.add_custom_button(__("暂停"), async () => {
				frm.__month_stop = true;
				await call("pause", frm.doc.name);
				await frm.reload_doc();
			});
		}
	};
})();

File: china_finance/services/month_end.py
"""Resumable month-end steps. Each request commits at most one source voucher."""

import json

import frappe
from frappe import _
from frappe.utils import get_first_day, get_last_day, getdate, now_datetime

from china_finance.services import voucher_preparation as preparation

RUN = "China Closing Run"
ACTIVE = ("Posting", "Checking", "Closing")
FIELDS = ("preparation_state", "preparation_data", "preparation_message")
ROLES = ("System Manager", "China Finance Manager")


def lock_company(company):
	frappe.db.sql("SELECT name FROM `tabCompany` WHERE name=%s FOR UPDATE", (company,))


def guard_period_write(doc, method=None):
	from china_finance.services.voucher import GL_SOURCE_DOCTYPES, get_posting_date

	if doc.doctype not in (
		*GL_SOURCE_DOCTYPES,
		"China Bank Receipt Import",
		"Bank Transaction",
	) or not doc.get("company"):
		return
	companies = {doc.company}
	old = doc.get_doc_before_save()
	if old and old.get("company"):
		companies.add(old.company)
	for company in sorted(companies):
		lock_company(company)
		if not frappe.db.has_column(RUN, "preparation_state"):
			continue
		active = frappe.db.sql(
			"""SELECT name, from_date, to_date FROM `tabChina Closing Run`
			WHERE company=%s AND docstatus=0 AND preparation_state IN ('Posting','Checking','Closing')
			FOR UPDATE""",
			company,
			as_dict=True,
		)
		for run in active:
			if getattr(frappe.flags, "china_month_end_run", None) == run.name:
				continue
			dates = [] if doc.doctype == "China Bank Receipt Import" else [get_posting_date(doc)]
			if doc.doctype == "Bank Transaction":
				dates = [getdate(doc.date)]
			if old:
				dates.append(
					getdate(old.date) if doc.doctype == "Bank Transaction" else get_posting_date(old)
				)
			if (
				doc.doctype == "China Bank Receipt Import"
				or not dates
				or any(getdate(run.from_date) <= date <= getdate(run.to_date) for date in dates)
			):
				frappe.throw(
					_("本期正在统一记账（{0}），请先在月末处理中暂停，再修改或导入。").format(run.name)
				)


def protect_run_fields(doc):
	old = doc.get_doc_before_save()
	for field in FIELDS:
		doc.set(field, old.get(field) if old else None)
	if old and old.get("preparation_state"):
		if any(doc.get(f) != old.get(f) for f in ("company", "from_date", "to_date", "closing_type")):
			frappe.throw(_("月末处理已开始，不能修改公司或期间，请新建正确期间的处理单"))


def _run(name):
	frappe.only_for(ROLES)
	doc = frappe.get_doc(RUN, name)
	doc.check_permission("write")
	preparation.check_company(doc.company)
	lock_company(doc.company)
	doc = frappe.get_doc(RUN, name, for_update=True)
	if doc.docstatus != 0:
		frappe.throw(_("该期间已经结账或作废"))
	if not preparation.mode_enabled(doc.company, doc.from_date):
		frappe.throw(_("请先启用凭证准备模式，且本期须在生效日期之后"))
	if (
		doc.closing_type != "Monthly"
		or getdate(doc.from_date) != get_first_day(doc.from_date)
		or getdate(doc.to_date) != get_last_day(doc.from_date)
	):
		frappe.throw(_("统一记账请选择完整自然月"))
	other = frappe.db.sql(
		"""SELECT name FROM `tabChina Closing Run` WHERE company=%s AND name!=%s
		AND docstatus=0 AND preparation_state IN ('Posting','Checking','Closing') FOR UPDATE""",
		(doc.company, doc.name),
	)
	if other:
		frappe.throw(_("公司还有进行中的月末处理：{0}").format(other[0][0]))
	return doc


def _store(doc, state, data, message):
	doc.db_set(
		{
			"preparation_state": state,
			"preparation_data": json.dumps(data, ensure_ascii=False, default=str),
			"preparation_message": message,
		}
	)
	return {
		"name": doc.name,
		"state": state,
		"message": message,
		"completed": len(data.get("done", [])),
		"total": len(data.get("items", [])),
		"period_closing_voucher": doc.period_closing_voucher,
		"checks": data.get("checks", []),
	}


def _drafts(doc):
	drafts = preparation.draft_documents(doc.company, doc.from_date, doc.to_date)
	count = frappe.db.count(
		"Journal Entry",
		{"company": doc.company, "docstatus": 0, "posting_date": ["between", [doc.from_date, doc.to_date]]},
	)
	if count != len(drafts):
		frappe.throw(_("没有读取本期全部凭证的权限，不能统一记账"), frappe.PermissionError)
	return drafts


def _preflight(doc):
	from china_finance.services.bank_receipt_import import pending_receipts
	from china_finance.services.closing import run_closing_checks

	blocked = pending_receipts(doc.company, doc.from_date, doc.to_date, allow_drafts=True)
	if blocked:
		frappe.throw(_("回单仍有待处理项：{0}").format("、".join(r.transaction_id for r in blocked[:20])))
	checks = run_closing_checks(
		doc.company, doc.from_date, doc.to_date, doc.period_closing_voucher, doc.closing_type
	)
	# These inputs are independent of the ledgers being posted. Other checks run again afterwards.
	config_codes = {"CONFIGURATION_READINESS", "VOUCHER_HASH"}
	bad = [
		r
		for r in checks
		if not r["passed"]
		and r["severity"] == "Blocking"
		and (r["check_code"] in config_codes or "TEMPLATE" in r["check_code"])
	]
	if bad:
		frappe.throw("；".join(r["description"] + "：" + r["details"] for r in bad))
	if frappe.db.count(
		"Journal Entry", {"company": doc.company, "docstatus": 0, "posting_date": ["<", doc.from_date]}
	):
		frappe.throw(_("存在更早期间的未记账凭证，请先核对并处理"))
	drafts = _drafts(doc)
	items = []
	for voucher in drafts:
		voucher.check_permission("submit")
		preparation.validate_draft(voucher)
		voucher.reload()
		if not preparation.is_ready(voucher):
			frappe.throw(_("{0} 尚未核对或内容已修改，请在查凭证中核对完成").format(voucher.name))
		items.append({"name": voucher.name, "hash": preparation.content_hash(voucher)})
	return items


@frappe.whitelist(methods=["POST"])
def start(name):
	doc = _run(name)
	data = json.loads(doc.preparation_data or "{}")
	if doc.preparation_state in ACTIVE:
		return _store(doc, doc.preparation_state, data, doc.preparation_message)
	point = "month_start"
	frappe.db.savepoint(point)
	items = _preflight(doc)
	# Never reuse a stale approved manifest after a pause. Already posted sources remain posted.
	done = []
	superseded = data.get("superseded", [])
	for item in data.get("done", []):
		state = frappe.db.get_value("Journal Entry", item["name"], "docstatus")
		if state == 2:
			superseded.append(item)
		elif state == 1:
			done.append(item)
		else:
			frappe.throw(_("原批次凭证 {0} 状态异常，请核对历史处理记录").format(item["name"]))
	data = {
		"items": [{"name": row["name"], "hash": row.get("hash")} for row in done] + items,
		"done": done,
		"superseded": superseded,
		"started_by": frappe.session.user,
		"started_on": str(now_datetime()),
	}
	return _store(doc, "Posting", data, _("已检查凭证范围，开始统一记账"))


@frappe.whitelist(methods=["POST"])
def pause(name):
	doc = _run(name)
	data = json.loads(doc.preparation_data or "{}")
	return _store(doc, "Paused", data, _("已暂停。已记账凭证保留，继续前将重新检查剩余草稿。"))


def _verify_posted(name):
	if frappe.db.get_value("Journal Entry", name, "docstatus") != 1 or not frappe.db.exists(
		"China Accounting Voucher",
		{"source_key": "Posting|Journal Entry|" + name, "docstatus": 1, "status": "Posted"},
	):
		frappe.throw(_("凭证 {0} 尚未记账或缺少有效会计快照").format(name))


def _post_one(doc, data):
	from china_finance.services.voucher import _complete_voucher_workflow

	item = data["items"][len(data["done"])]
	voucher = frappe.get_doc("Journal Entry", item["name"], for_update=True)
	voucher.check_permission("submit")
	if (
		voucher.docstatus != 0
		or preparation.content_hash(voucher) != item["hash"]
		or not preparation.is_ready(voucher)
	):
		frappe.throw(_("{0} 状态或核对版本已变化，请暂停后重新检查").format(voucher.name))
	_complete_voucher_workflow(voucher)
	_verify_posted(voucher.name)
	preparation.promote_cash_plan(voucher, confirm=False)
	voucher.add_comment("Comment", _("月末统一记账：{0}").format(doc.name))
	data["done"].append(item)
	return _store(doc, "Posting", data, _("已记账 {0}").format(voucher.name))


def _after_posting(doc, data):
	from china_finance.services.bank_receipt_import import receipt_status, reconcile_receipt
	from china_finance.services.cash_flow_assignment import get_assignment_coverage
	from china_finance.services.closing import create_period_closing_voucher

	if _drafts(doc):
		frappe.throw(_("本期仍有未记账草稿，请暂停后重新检查"))
	for item in data["done"]:
		_verify_posted(item["name"])
		preparation.promote_cash_plan(frappe.get_doc("Journal Entry", item["name"]))
	for row in frappe.get_list(
		"China Bank Receipt",
		filters={"company": doc.company, "posting_date": ["between", [doc.from_date, doc.to_date]]},
		pluck="name",
		limit_page_length=0,
	):
		receipt = frappe.get_doc("China Bank Receipt", row)
		if receipt_status(receipt) == "已关联待核销":
			reconcile_receipt(receipt.name)
	coverage = get_assignment_coverage(doc.company, doc.from_date, doc.to_date)
	if not coverage["passed"]:
		frappe.throw(_("现金流项目需要核对：{0}").format(coverage["details"]))
	if (
		doc.period_closing_voucher
		and frappe.db.get_value("Period Closing Voucher", doc.period_closing_voucher, "docstatus") == 2
	):
		doc.db_set("period_closing_voucher", None)
	if not doc.period_closing_voucher:
		result = create_period_closing_voucher(
			doc.company, doc.from_date, doc.to_date, allow_first_period=True
		)
		doc.db_set("period_closing_voucher", result["name"])
	return _store(doc, "Closing", data, _("凭证已记账，开始损益结转"))


def _close_profit(doc, data):
	from china_finance.services.closing import run_closing_checks, save_and_complete_period_closing_voucher
	from china_finance.services.voucher import create_voucher_from_source

	pcv = frappe.get_doc("Period Closing Voucher", doc.period_closing_voucher, for_update=True)
	pcv.check_permission("submit")
	if pcv.docstatus == 0:
		from frappe.model.workflow import get_workflow_name

		if get_workflow_name(pcv.doctype):
			save_and_complete_period_closing_voucher(pcv.name)
		else:
			pcv.submit()
		return _store(doc, "Closing", data, _("损益结转已提交，等待总账处理完成"))
	pcv.reload()
	if pcv.docstatus != 1 or pcv.gle_processing_status == "Failed":
		frappe.throw(_("损益结转失败，请打开结转凭证处理错误后继续"))
	if pcv.gle_processing_status != "Completed":
		return _store(doc, "Closing", data, _("损益结转总账处理中，请稍后继续"))
	create_voucher_from_source(pcv)
	checks = run_closing_checks(doc.company, doc.from_date, doc.to_date, pcv.name, doc.closing_type)
	data["checks"] = checks
	bad = [r for r in checks if r["severity"] == "Blocking" and not r["passed"]]
	if bad:
		return _store(doc, "Paused", data, _("结账检查未通过，请处理下方问题后继续"))
	return _store(doc, "Ready", data, _("记账与结转完成，请核对正式报表后确认结账"))


@frappe.whitelist(methods=["POST"])
def advance(name):
	doc = _run(name)
	data = json.loads(doc.preparation_data or "{}")
	if doc.preparation_state not in ACTIVE:
		frappe.throw(_("请先检查并开始，或继续上次处理"))
	point = "month_step"
	frappe.db.savepoint(point)
	previous = getattr(frappe.flags, "china_month_end_run", None)
	frappe.flags.china_month_end_run = doc.name
	try:
		if doc.preparation_state == "Posting":
			if len(data.get("done", [])) < len(data.get("items", [])):
				return _post_one(doc, data)
			return _store(doc, "Checking", data, _("凭证记账完成，检查核销和现金流"))
		if doc.preparation_state == "Checking":
			return _after_posting(doc, data)
		return _close_profit(doc, data)
	except Exception as exc:
		frappe.db.rollback(save_point=point)
		doc.reload()
		data = json.loads(doc.preparation_data or "{}")
		return _store(doc, "Failed", data, str(exc))
	finally:
		frappe.flags.china_month_end_run = previous


@frappe.whitelist(methods=["POST"])
def finish(name):
	from china_finance.services.closing import submit_closing_run

	frappe.only_for(ROLES)
	existing = frappe.get_doc(RUN, name)
	existing.check_permission("submit")
	preparation.check_company(existing.company)
	if existing.docstatus == 1 and existing.status == "Closed":
		return {
			"name": name,
			"status": "Closed",
			"already_submitted": True,
			"archive_package": existing.archive_package,
		}
	doc = _run(name)
	if doc.preparation_state != "Ready":
		frappe.throw(_("请先完成统一记账和损益结转"))
	if _drafts(doc):
		frappe.throw(_("本期仍有未记账凭证"))
	if (
		frappe.db.get_value("Period Closing Voucher", doc.period_closing_voucher, "gle_processing_status")
		!= "Completed"
	):
		frappe.throw(_("损益结转总账尚未完成"))
	result = submit_closing_run(name)
	return result


@frappe.whitelist(methods=["POST"])
def retry_archive(name):
	frappe.only_for(ROLES)
	doc = frappe.get_doc(RUN, name)
	doc.check_permission("submit")
	preparation.check_company(doc.company)
	if doc.docstatus != 1 or doc.status != "Closed":
		frappe.throw(_("请先完成结账"))
	if doc.archive_package:
		return {"archive_package": doc.archive_package}
	frappe.enqueue(
		"china_finance.services.closing.create_closing_archive",
		closing_run_name=name,
		queue="short",
		enqueue_after_commit=True,
		job_id="china-closing-archive:" + name,
		deduplicate=True,
	)
	return {"message": _("归档任务已加入队列，请稍后刷新")}
